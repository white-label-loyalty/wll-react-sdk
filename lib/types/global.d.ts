// Global type definitions for React Native environment

declare const __DEV__: boolean;
