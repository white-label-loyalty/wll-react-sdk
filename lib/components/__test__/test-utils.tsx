import { render } from '@testing-library/react';
import React from 'react';
import { CoreApiUrl, WllSdkProvider } from '../../context/WllSdkContext';
import { defaultTheme } from '../../utils/styling';

const defaultConfig = {
  apiKey: 'test-api-key',
  userToken: 'test-user-token',
  coreApiUrl: CoreApiUrl.STAGING_EU,
};

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <WllSdkProvider
      config={defaultConfig}
      theme={defaultTheme}
      navigationConfig={{}}
    >
      {children}
    </WllSdkProvider>
  );
};

const customRender = (ui: React.ReactElement, options = {}) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
