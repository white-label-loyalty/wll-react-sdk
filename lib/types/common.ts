import { ImageProps, ViewStyle } from 'react-native';

export type ImagePropsNoSource = Omit<ImageProps, 'source'>;

export type FlexDirection = 'row' | 'row-reverse' | 'column' | 'column-reverse';

export type Justify =
  | 'start'
  | 'end'
  | 'center'
  | 'between'
  | 'around'
  | 'evenly';

export type Align = 'start' | 'end' | 'center' | 'stretch' | 'baseline';

export type LayoutProps = {
  children: React.ReactNode;
  justify?: Justify;
  align?: Align;
  direction?: FlexDirection;
  style?: ViewStyle;
};

export const justifyMap: Record<Justify, ViewStyle['justifyContent']> = {
  start: 'flex-start',
  end: 'flex-end',
  center: 'center',
  between: 'space-between',
  around: 'space-around',
  evenly: 'space-evenly',
} as const;

export const alignMap: Record<Align, ViewStyle['alignItems']> = {
  start: 'flex-start',
  end: 'flex-end',
  center: 'center',
  stretch: 'stretch',
  baseline: 'baseline',
} as const;
