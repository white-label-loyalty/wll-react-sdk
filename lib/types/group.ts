import { TSection } from './section';

export type TGroup = {
  name: string;
  active: boolean;
  id: string;
  createdAt: string;
  updatedAt: string;
  sections: TSection[];
};
