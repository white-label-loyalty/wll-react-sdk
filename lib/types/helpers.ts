/**
 * Makes a configuration type with required and optional fields,
 * providing default values for optional fields.
 */
export type SafeConfig<T, R extends keyof T> = Required<Pick<T, R>> &
  Partial<Omit<T, R>>;

/**
 * Adds optional children to any component props.
 */
export type WithChildren<T = {}> = T & { children?: React.ReactNode };
