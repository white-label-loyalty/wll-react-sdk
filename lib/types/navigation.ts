export type NavigationType = 'external' | 'internal' | 'modal';

type NavigationHandlerParams = {
  target: string;
  windowTarget: string;
};

export type NavigationConfig = {
  navigationHandlers?: {
    external?: (params: NavigationHandlerParams) => Promise<void>;
    internal?: (params: NavigationHandlerParams) => Promise<void>;
    modal?: (params: NavigationHandlerParams) => Promise<void>;
  };
  external?: (params: NavigationHandlerParams) => Promise<void>;
  internal?: (params: NavigationHandlerParams) => Promise<void>;
  modal?: (params: NavigationHandlerParams) => Promise<void>;
  baseUrl?: string;
};
