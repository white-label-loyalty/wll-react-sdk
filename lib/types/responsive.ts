export type DimensionMode = 'desktop' | 'tablet' | 'mobile';

export type WindowDimensions = {
  width: number;
  height: number;
};

export type ResponsiveContextType = {
  dimensionMode: DimensionMode;
  isTablet: boolean;
  isMobile: boolean;
  isDesktop: boolean;
};
