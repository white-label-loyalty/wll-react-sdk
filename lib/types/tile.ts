export enum CtaAction {
  sameWindow = 'SAME_WINDOW',
  newWindow = 'NEW_WINDOW',
}

export enum BadgeTileType {
  latestEarned = 'LATEST_EARNED',
  specificBadge = 'SPECIFIC_BADGE',
}

export enum TierTileType {
  currentTier = 'CURRENT_TIER',
  nextTier = 'NEXT_TIER',
  specificTier = 'SPECIFIC_TIER',
}

export enum TileType {
  Banner = 'BANNER',
  Points = 'POINTS',
  Content = 'CONTENT',
  Reward = 'REWARD',
  Badge = 'BADGE',
  RewardCategory = 'REWARD_CATEGORY',
  Tier = 'TIER',
}

export enum TileHeight {
  Half = 'HALF',
  Full = 'FULL',
}
type Badge = {
  id: string | null;
  name: string | null;
  description: string | null;
  artworkUrl: string | null;
  createdAt: string | null;
  updatedAt: string | null;
};
export class RewardCategory {
  name?: string | null;
  priority?: number;
  type?: string | null;
  id?: string | null;
  createdAt?: string | null;
  updatedAt?: string | null;
  description?: string | null;
  metadata?: string | null;
  pictureUrl?: string | null;
  rewards?: Reward[];
  parent?: string | null;
}

export class Reward {
  id?: string | null;
  createdAt?: string | null;
  updatedAt?: string | null;
  name?: string | null;
  pictureUrl?: string | null;
  description?: string | null;
  value?: number | null;
  price?: number;
  priority?: number;
  discounts?: any[] | null;
  redemptionChannels?: string[] | null;
  venues?: any[] | null;
  purchasableForAudiences?: string[] | null;
  category?: string | null;
  terms?: string | null;
  availability?: Availability;
  purchasable?: boolean;
  tier?: string | null;
  summary?: string | null;
  redemptionMessage?: string | null;
  visibilityCriteria?: string | null;
  logoUrl?: string | null;
  type?: string | null;
  codeType?: string | null;
  code?: string | null;
  purchaseExpiration?: string | null;
  hideCode?: boolean;
  notificationConfig?: string | null;
}
type Availability = {
  start: string | null;
  end: string | null;
};
type Effectivity = {
  start: string | null;
  end: string | null;
};

export type TierType = {
  id: string | null;
  name: string | null;
  description: string | null;
  artworkUrl: string | null;
  priority: number;
  pointsRequirement: number;
  calculation: string | null;
  accumulationPeriod: string | null;
  effectivity: Effectivity;
  createdAt: string | null;
  updatedAt: string | null;
  earnedPoints: number;
  attained: boolean;
};

export class BannerTileConfig {
  imageUrl?: string | null;
  title?: string | null;
  description?: string | null;
  ctaText?: string | null;
  ctaLink?: string | null;
  ctaAction?: CtaAction;
}

export class PointsTileConfig {
  title?: string | null;
  multiplier?: number;
  prefix?: string | null;
  suffix?: string | null;
  imageUrl?: string | null;
  points?: number;
}

export class ContentTileConfig {
  title?: string | null;
  subtitle?: string | null;
  imageUrl?: string | null;
}

export class RewardTileConfig {
  reward?: Reward;
  showPrice?: boolean;
}

export class BadgeTileConfig {
  badgeTileType?: BadgeTileType;
  badgeId?: string | null;
  badge?: Badge;
}

export class RewardCategoryTileConfig {
  categoryId?: string | null;
  allowDecorationOverlay?: boolean;
  rewardCategory?: RewardCategory;
}

export class TierTileConfig {
  tierId?: string | null;
  tier?: TierType;
  type?: TierTileType;
}

export type Tile = {
  id: string | null;
  type: TileType;
  active: boolean;
  createdAt: string | null;
  updatedAt: string | null;
  visibilityCriteria: object | null;
  tileHeight: TileHeight;
  configuration: TileConfig;
};

const getConfigForTileType = (tileType: TileType) => {
  switch (tileType) {
    case TileType.Banner:
      return BannerTileConfig;
    case TileType.Points:
      return PointsTileConfig;
    case TileType.Content:
      return ContentTileConfig;
    case TileType.Reward:
      return RewardTileConfig;
    case TileType.Badge:
      return BadgeTileConfig;
    case TileType.RewardCategory:
      return RewardCategoryTileConfig;
    case TileType.Tier:
      return TierTileConfig;
    default:
      throw new Error(`Tile type ${tileType} is not supported`);
  }
};

export type TileConfig = InstanceType<
  ReturnType<typeof getConfigForTileType>
>;
