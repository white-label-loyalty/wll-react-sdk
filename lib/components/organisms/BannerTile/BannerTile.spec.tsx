import '@testing-library/jest-dom';
import React from 'react';
import { CTALinkTarget, Tile, TileHeight, TileType } from '../../../types/tile';
import { render } from '../../__test__/test-utils';
import { BannerTile } from './index';

const BannerImageOnlyMock: Tile = {
  id: '3',
  type: TileType.Banner,
  active: true,
  createdAt: '',
  updatedAt: '',
  tileHeight: TileHeight.Full,
  priority: 1,
  configuration: {
    artworkUrl:
      'https://images.pexels.com/photos/6250883/pexels-photo-6250883.jpeg',
    title: '',
    description: '',
    ctaText: '',
    ctaLink: '',
    ctaLinkTarget: CTALinkTarget.sameWindow,
  },
};

const BannerTileMock = {
  ...BannerImageOnlyMock,
  configuration: {
    title: 'Summer Points Splash',
    ctaLink: '/summer-rewards',
    ctaText: 'Start Earning',
    artworkUrl:
      'https://images.pexels.com/photos/1209611/pexels-photo-1209611.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=2',
    description:
      'Double points on all summer essentials! Plus, unlock exclusive seasonal rewards when you spend $100 or more.',
    ctaLinkTarget: CTALinkTarget.sameWindow,
  },
};

describe('<BannerTile  />', () => {
  it('renders and matches snapshot', () => {
    const { container } = render(<BannerTile tile={BannerTileMock} />);
    expect(container).toMatchSnapshot();
  });
});

describe('<BannerTile /> Rendering States', () => {
  it('returns null when tile is not active', () => {
    const inactiveTile = {
      ...BannerTileMock,
      active: false,
    };
    const { container } = render(<BannerTile tile={inactiveTile} />);
    expect(container).toBeEmptyDOMElement();
  });
});

describe('<BannerTile /> Media', () => {
  it('renders artwork only when only artworkUrl is provided', () => {
    const { getByTestId, queryByTestId } = render(
      <BannerTile tile={BannerImageOnlyMock} />
    );
    expect(getByTestId('banner-tile-media')).toBeInTheDocument();
    expect(queryByTestId('banner-tile-title')).not.toBeInTheDocument();
    expect(queryByTestId('banner-tile-description')).not.toBeInTheDocument();
    expect(queryByTestId('banner-tile-cta')).not.toBeInTheDocument();
  });
});

describe('<BannerTile /> Other Components', () => {
  it('renders cta when ctaText and ctaLink are provided', () => {
    const { getByTestId } = render(<BannerTile tile={BannerTileMock} />);
    expect(getByTestId('banner-tile-cta')).toBeInTheDocument();
  });

  it('renders title and description when title and description are provided', () => {
    const { getByTestId } = render(<BannerTile tile={BannerTileMock} />);
    expect(getByTestId('banner-tile-title')).toBeInTheDocument();
    expect(getByTestId('banner-tile-description')).toBeInTheDocument();
  });
});

describe('<BannerTile /> Content', () => {
  it('renders correct content in title and description', () => {
    const { getByText } = render(<BannerTile tile={BannerTileMock} />);
    expect(getByText('Summer Points Splash')).toBeInTheDocument();
    expect(
      getByText(
        'Double points on all summer essentials! Plus, unlock exclusive seasonal rewards when you spend $100 or more.'
      )
    ).toBeInTheDocument();
  });

  it('renders CTA with correct text', () => {
    const { getByRole } = render(<BannerTile tile={BannerTileMock} />);
    const button = getByRole('button', { name: 'Start Earning' });
    expect(button).toBeInTheDocument();
  });
});

describe('<BannerTile /> Edge Cases', () => {
  const emptyBannerMock = {
    ...BannerImageOnlyMock,
    configuration: {
      ...BannerImageOnlyMock.configuration,
      artworkUrl: '',
    },
  };

  it('handles missing artworkUrl gracefully', () => {
    const { queryByTestId } = render(<BannerTile tile={emptyBannerMock} />);
    expect(queryByTestId('banner-tile-media')).not.toBeInTheDocument();
  });

  it('handles missing optional props gracefully', () => {
    const partialBannerMock = {
      ...BannerTileMock,
      configuration: {
        ...BannerTileMock.configuration,
        description: undefined,
        ctaText: undefined,
        ctaLink: undefined,
      },
    };
    const { queryByTestId } = render(<BannerTile tile={partialBannerMock} />);
    expect(queryByTestId('banner-tile-description')).not.toBeInTheDocument();
    expect(queryByTestId('banner-tile-cta')).not.toBeInTheDocument();
  });
});
