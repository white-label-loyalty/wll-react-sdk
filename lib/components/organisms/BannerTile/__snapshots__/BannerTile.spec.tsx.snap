// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BannerTile  /> renders and matches snapshot 1`] = `
<div>
  <article
    aria-label="Summer Points Splash"
    class="css-view-175oi2r r-alignItems-1awozwy r-flexDirection-18u37iz r-justifyContent-1wtj0ep r-maxWidth-q90dru r-overflow-1udh08x r-width-13qz1uu"
    data-testid="banner-tile"
    role="article"
    style="background-color: rgb(248, 247, 252); border-top-left-radius: 20px; border-top-right-radius: 20px; border-bottom-right-radius: 20px; border-bottom-left-radius: 20px;"
  >
    <div
      aria-label="Banner image for Summer Points Splash"
      class="css-view-175oi2r r-aspectRatio-1ujkv8a r-marginRight-at4v1u r-maxHeight-fu70me r-minHeight-1rd2zbf r-overflow-1udh08x r-position-bnwqim"
      data-testid="banner-tile-media"
      role="img"
      style="width: 30%;"
    >
      <div
        class="css-view-175oi2r r-alignItems-1awozwy r-justifyContent-1777fci r-height-1pi2tsx r-objectFit-1t7uo4s r-position-u8s1d r-width-13qz1uu"
        style="background-color: rgba(57, 46, 215, 0.2);"
      >
        <div
          class="css-view-175oi2r r-bottom-1p0dtai r-left-1d2f490 r-position-u8s1d r-right-zchlnj r-top-ipm5af"
          style="background-color: rgba(57, 46, 215, 0.2); opacity: 1;"
        />
        <div
          aria-label="Banner image for Summer Points Splash"
          class="css-view-175oi2r r-flexBasis-1mlwlqe r-overflow-1udh08x r-zIndex-417010 r-bottom-1p0dtai r-left-1d2f490 r-position-u8s1d r-right-zchlnj r-top-ipm5af"
          role="img"
          style="opacity: 0;"
        >
          <div
            class="css-view-175oi2r r-backgroundColor-1niwhzg r-backgroundPosition-vvn4in r-backgroundRepeat-u6sd8q r-bottom-1p0dtai r-height-1pi2tsx r-left-1d2f490 r-position-u8s1d r-right-zchlnj r-top-ipm5af r-width-13qz1uu r-zIndex-1wyyakw r-backgroundSize-4gszlv"
            style="background-image: url(https://images.pexels.com/photos/1209611/pexels-photo-1209611.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=2);"
          />
          <img
            alt="Banner image for Summer Points Splash"
            class="css-accessibilityImage-9pa8cd"
            draggable="false"
            src="https://images.pexels.com/photos/1209611/pexels-photo-1209611.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=2"
          />
        </div>
      </div>
    </div>
    <div
      class="css-view-175oi2r"
      style="flex: 1;"
    >
      <h1
        aria-label="Summer Points Splash"
        class="css-text-146c3p1 r-fontFamily-lhg2x6 r-fontSize-1yjpyg1 r-fontWeight-b88u0q r-marginBottom-1a8r3js"
        data-testid="banner-tile-title"
        dir="auto"
        role="heading"
        style="color: rgb(0, 0, 0);"
      >
        Summer Points Splash
      </h1>
      <div
        class="css-text-146c3p1 r-fontFamily-lhg2x6 r-color-1khnkhu r-fontSize-1inkyih r-marginBottom-10qg9sp"
        data-testid="banner-tile-description"
        dir="auto"
      >
        Double points on all summer essentials! Plus, unlock exclusive seasonal rewards when you spend $100 or more.
      </div>
      <button
        aria-label="Start Earning"
        class="css-view-175oi2r r-cursor-1loqt21 r-touchAction-1otgn73 r-alignItems-1awozwy r-alignSelf-k200y r-justifyContent-1777fci r-paddingBlock-1mmae3n r-paddingInline-10tapur"
        data-testid="banner-tile-cta"
        role="button"
        style="background-color: rgb(255, 106, 61); border-top-left-radius: 9px; border-top-right-radius: 9px; border-bottom-right-radius: 9px; border-bottom-left-radius: 9px;"
        tabindex="0"
        type="button"
      >
        <div
          class="css-text-146c3p1 r-fontSize-1i10wst r-fontWeight-b88u0q r-textAlign-q4m81j r-textTransform-tsynxw"
          dir="auto"
          style="color: rgb(255, 255, 255);"
        >
          Start Earning
        </div>
      </button>
    </div>
  </article>
</div>
`;
